File: ios/PaddleupShared/Sources/PaddleupShared/StrokeDetector.swift
import Foundation

/// Real-time stroke detector. Mirrors the Kotlin implementation byte-for-byte
/// so the same sensor stream produces the same stroke times on iOS and Android.
public final class StrokeDetector {
    public struct Stroke: Sendable {
        public let tSec: Double
        public let rateSpm: Double
        public let confidence: Double
    }

    private let sampleRateHz: Double
    private let minStrokeRateSpm: Double
    private let maxStrokeRateSpm: Double
    private let peakThreshold: Double
    private let refractorySec: Double

    private let hpAlpha = 0.97
    private let lpAlpha = 0.25

    private var hpPrevIn = 0.0
    private var hpPrevOut = 0.0
    private var lpPrev = 0.0

    private var lastStrokeT: Double = -1
    private var lastSampleT: Double = -1
    private var lastValue: Double = 0
    private var rising = false

    public init(
        sampleRateHz: Double = 50,
        minStrokeRateSpm: Double = 30,
        maxStrokeRateSpm: Double = 120,
        peakThreshold: Double = 0.6
    ) {
        self.sampleRateHz = sampleRateHz
        self.minStrokeRateSpm = minStrokeRateSpm
        self.maxStrokeRateSpm = maxStrokeRateSpm
        self.peakThreshold = peakThreshold
        self.refractorySec = 60.0 / maxStrokeRateSpm
    }

    /// Feed one sample of linear acceleration (gravity already removed).
    public func onSample(tSec: Double, ax: Double, ay: Double, az: Double) -> Stroke? {
        let magnitude = sqrt(ax * ax + ay * ay + az * az)

        let hp = hpAlpha * (hpPrevOut + magnitude - hpPrevIn)
        hpPrevIn = magnitude
        hpPrevOut = hp

        let v = lpPrev + lpAlpha * (hp - lpPrev)
        lpPrev = v

        var stroke: Stroke?
        if lastSampleT > 0 {
            let nowRising = v > lastValue
            if rising && !nowRising && lastValue > peakThreshold {
                if lastStrokeT < 0 || (tSec - lastStrokeT) >= refractorySec {
                    let rate = lastStrokeT < 0 ? 0 : 60.0 / (tSec - lastStrokeT)
                    if rate == 0 || (rate >= minStrokeRateSpm && rate <= maxStrokeRateSpm) {
                        let prom = lastValue - peakThreshold
                        let conf = max(0, min(1, prom / (prom + 0.5)))
                        stroke = Stroke(tSec: lastSampleT, rateSpm: rate, confidence: conf)
                        lastStrokeT = lastSampleT
                    }
                }
            }
            rising = nowRising
        }
        lastValue = v
        lastSampleT = tSec
        return stroke
    }

    public func reset() {
        hpPrevIn = 0; hpPrevOut = 0; lpPrev = 0
        lastStrokeT = -1; lastSampleT = -1; lastValue = 0
        rising = false
    }
}

File: src/services/stroke-detector.ts
/**
 * Real-time stroke detector for outrigger paddling. Same algorithm we'll
 * use on the watch (when we add it natively), so results are consistent.
 *
 * Pipeline: |a| → high-pass → low-pass → peak pick with refractory window.
 * Designed to be deterministic so the same accelerometer stream produces
 * the same stroke times across runs.
 */
export interface Stroke {
  tSec: number;
  rateSpm: number;
  confidence: number;
}

export interface StrokeDetectorOptions {
  sampleRateHz?: number;
  minStrokeRateSpm?: number;
  maxStrokeRateSpm?: number;
  peakThreshold?: number;
}

export class StrokeDetector {
  private readonly minSpm: number;
  private readonly maxSpm: number;
  private readonly peakThreshold: number;
  private readonly refractorySec: number;

  private readonly hpAlpha = 0.97;
  private readonly lpAlpha = 0.25;

  private hpPrevIn = 0;
  private hpPrevOut = 0;
  private lpPrev = 0;

  private lastStrokeT = -1;
  private lastSampleT = -1;
  private lastValue = 0;
  private rising = false;

  constructor(opts: StrokeDetectorOptions = {}) {
    this.minSpm = opts.minStrokeRateSpm ?? 30;
    this.maxSpm = opts.maxStrokeRateSpm ?? 120;
    this.peakThreshold = opts.peakThreshold ?? 0.6;
    this.refractorySec = 60 / this.maxSpm;
  }

  /**
   * Feed one sample of linear acceleration (gravity already removed).
   * Returns a Stroke if a peak was detected at this sample, else null.
   */
  onSample(tSec: number, ax: number, ay: number, az: number): Stroke | null {
    const magnitude = Math.sqrt(ax * ax + ay * ay + az * az);

    const hp = this.hpAlpha * (this.hpPrevOut + magnitude - this.hpPrevIn);
    this.hpPrevIn = magnitude;
    this.hpPrevOut = hp;

    const v = this.lpPrev + this.lpAlpha * (hp - this.lpPrev);
    this.lpPrev = v;

    let stroke: Stroke | null = null;
    if (this.lastSampleT > 0) {
      const nowRising = v > this.lastValue;
      if (this.rising && !nowRising && this.lastValue > this.peakThreshold) {
        // The refractory window is the only thing that rejects a peak. A peak
        // past it IS a stroke; the rate band decides whether the INTERVAL is a
        // believable cadence, not whether the stroke happened.
        //
        // This used to gate the stroke itself on the rate band, and only
        // advanced lastStrokeT when a stroke was accepted — which made the
        // detector jam permanently. One gap longer than 60/minSpm (2s at the
        // default) pushed the implied rate below minSpm, so the peak was
        // rejected AND lastStrokeT stayed put; every later peak was measured
        // from that same frozen point, so the rate only fell further and every
        // stroke for the rest of the session was rejected. A single pause —
        // pushing off, a rest, a wave — killed stroke counting for good. It is
        // why an 85-minute paddle came back with a stroke count of 1.
        if (this.lastStrokeT < 0 || tSec - this.lastStrokeT >= this.refractorySec) {
          const raw = this.lastStrokeT < 0 ? 0 : 60 / (tSec - this.lastStrokeT);
          // Out-of-band interval means the paddler stopped and started again:
          // still a stroke, but it begins a new cadence measurement rather than
          // reporting an implausible one. 0 reads as "unknown" downstream and
          // is excluded from averages.
          const rate = raw >= this.minSpm && raw <= this.maxSpm ? raw : 0;
          const prom = this.lastValue - this.peakThreshold;
          const conf = Math.max(0, Math.min(1, prom / (prom + 0.5)));
          stroke = { tSec: this.lastSampleT, rateSpm: rate, confidence: conf };
          // ALWAYS advance, so cadence re-anchors after any gap.
          this.lastStrokeT = this.lastSampleT;
        }
      }
      this.rising = nowRising;
    }
    this.lastValue = v;
    this.lastSampleT = tSec;
    return stroke;
  }

  reset(): void {
    this.hpPrevIn = 0;
    this.hpPrevOut = 0;
    this.lpPrev = 0;
    this.lastStrokeT = -1;
    this.lastSampleT = -1;
    this.lastValue = 0;
    this.rising = false;
  }
}
